src/Repulsion.cc

```diff
@@ -0,0 +1,354 @@
+
+    if (rootnode == nullptr) {
+      return frags;
+    }
+
+    std::vector<Vector3d> fs(frags.size(), {0.0, 0.0, 0.0});
+
+    GetForces(rootnode, rootnode, maps, fs);
+
+    double temp_timedelta = delta_time;
+    auto adaptive_tm = GetAdaptiveTimeDelta(frags, fs);
+    if ((adaptive_tm < temp_timedelta) && (totalTime / adaptive_tm < (2 * iterations))) {
+      temp_timedelta = adaptive_tm;
+    }
+
+    auto delta_x = Iterate(frags, fs, temp_timedelta, frag_masses);
+
+    for (int i = 0; i < nucleons.size(); i++) {
+      nucleons[i].x += delta_x[maps[i]].x;
+      nucleons[i].y += delta_x[maps[i]].y;
+      nucleons[i].z += delta_x[maps[i]].z;
+    }
+    time += temp_timedelta;
+  }
+
+  for (auto& frag : frags) {
+    frag->SetMomentum(frag->GetMomentum() / MeVc);
+  }
+  return frags;
+}
+
+std::vector<Vector3d> Iterate(G4FragmentVector& frags, const std::vector<Vector3d>& fs, double time_delta, const std::vector<double>& frag_masses) {
+  std::vector<Vector3d> delta_x(frags.size());
+
+  for (size_t i = 0; i < frags.size(); ++i) {
+    if (frags[i]->GetZ_asInt() == 0) {
+      continue;
+    }
+
+    G4LorentzVector currentMomentum = frags[i]->GetMomentum();
+    G4ThreeVector spatialMomentum = currentMomentum.vect();
+
+    G4ThreeVector halfstepMomentumIncrement(
+        0.5 * time_delta * (fs[i].x),
+        0.5 * time_delta * (fs[i].y),
+        0.5 * time_delta * (fs[i].z)
+    );
+
+    G4ThreeVector newMomentum = spatialMomentum + halfstepMomentumIncrement;
+    double gamma = GetGamma(currentMomentum, frag_masses[i]);
+    G4ThreeVector velocity = newMomentum / (frag_masses[i] * gamma);
+
+    delta_x[i] = Vector3d(time_delta * velocity.x(),
+                            time_delta * velocity.y(),
+                            time_delta * velocity.z());
+
+    newMomentum = newMomentum + halfstepMomentumIncrement;
+    G4LorentzVector finalMomentum(std::sqrt(newMomentum.mag() * newMomentum.mag() + std::pow(frag_masses[i] * speed_light, 2)), newMomentum);
+
+    frags[i]->SetMomentum(finalMomentum);
+  }
+  return delta_x;
+}
+
+void LazyGetForces(const aamcc::NucleonVector* nucleons, const std::vector<int>& maps, std::vector<Vector3d>& fs) {
+  for (size_t i = 0; i < nucleons->size(); i++) {
+    for (size_t j = i + 1; j < nucleons->size(); j++) {
+      auto nuc = (*nucleons)[i];
+      auto nuc2 = (*nucleons)[j];
+      if (nuc.isospin != 1 || nuc2.isospin != 1) {
+        continue;
+      }
+      Vector3d r2 = {nuc2.GetX(), nuc2.GetY(), nuc2.GetZ()};
+      Vector3d r = {nuc.GetX(), nuc.GetY(), nuc.GetZ()};
+      Vector3d tempf = DuoForce(r2, r, 1);
+      fs[maps[i]] += tempf;
+      fs[maps[j]] -= tempf;
+    }
+  }
+}
+
+std::vector<double> GetFragmentMasses(G4FragmentVector& fragments) {
+  std::vector<double> ms;
+  for (auto& frag : fragments) {
+    double mass = (frag->GetZ_asInt() * p_mass) + (frag->GetA_asInt() - frag->GetZ_asInt()) * n_mass;
+    frag->SetMomentum(MeVc * frag->GetMomentum());
```

Comment:
I've traced that you call this function only once, but side effects like these are nasty. Please, make momentum scaling a separate function, or at least rename this one to make it apparent.